design-proposals/klip-4-produce-data.md

```diff
@@ -0,0 +1,277 @@
+  * Do not require `PARTITIONS`/`REPLICAS` to create the kafka topic. This is a slight change that
+  will make it easier to experiment, but may open the door to accidentally creating kafka topics.
+  * Introduce another `CREATE TOPIC` API which will create empty topics. I believe this will cause
+  confusion, as it departs farther from SQL standard and puts `TOPIC` on the same level as the
+  duality of Tables/Streams (which it is not, it is a physical manifestation of a stream/table).
+  * Don't fail when creating streams/topics with non-existing topics, and create those topics if
+  a call to `INSERT INTO ...` is called with a stream/table target that does not exist. This seems
+  more "magical" and runs into issues when specifying partition/replica counts.
+
+### Code Enhancements
+
+KSQL in interactive mode is already configured to produce data to a command topic. This KLIP can
+leverage that existing functionality, which will require piping through various configurations into
+the `ServiceContext`. Beyond that, we must introduce a new command and Statement type to reflect the
+new functionality. 
+
+In interactive mode, this will be handled without distributing the statement to other servers. This
+operation will not be available in headless mode for the first version of the feature.
+
+### Schema Validation
+
+`INSERT INTO ... VALUES` will only support inserting into values for existing KSQL Streams/Tables,
+which already have a schema associated with them. Before producing a record to the underlying topic,
+the values for the fields will be verified against the schema. If the columns contain `ROWKEY`, the
+value will be verified against the key schema but will not populate the corresponding field in the 
+value. Populating columns with `ROWTIME` will not be supported.
+
+### Serialization Support
+
+JSON, AVRO and DELIMITED value formats will all be supported, the latter of which will not accept
+complex types. Since we know the schema of the inserts, we can initially load the values into a 
+connect data type. We already have the proper serializers to convert from connect to each of the
+supported serialization formats.
+
+## Key Handling
+
+The underlying data and the way it is represented in KSQL sometimes differ, and this `INSERT INTO`
+functionality needs to bridge that gap. For example take the following KSQL statement:
+```sql
+CREATE TABLE bar (id VARCHAR, foo INT) WITH (key="id")
+```
+
+In this case, we expect kafka records that resemble 
+```json
+{"key": "bar", "value": {"id": "bar", "foo": 123}}
+```
+
+If this is not the case (e.g. the value of `"key"` and of `"value"->"id"` do not match), then the
+KSQL behavior is not well defined. To all for all levels of flexibility, `INSERT INTO` should allow
+for all cases that kafka supports. Therefore, `ROWKEY` will always be available in the list of
+column names as the first column. The following examples would all be possible: 
```

Comment:
> Therefore, `ROWKEY` will always be available in the list of column names as the first column.

I don't see the benefit of this. I think we should use the `KEY` from the `WITH` and set the key ourselves. 

`INSERT INTO` is not a replacement for `kafkacat` in which people can construct whatever they want in a Kafka message. It should be an easily-accessible, safe, way of populating data in KSQL.

Reply:
>  It should be an easily-accessible, safe, way of populating data in KSQL.

This makes a lot of sense. Please see the updated section "ROWKEY and ROWTIME handling" in my most recent update.

Reply:
I've not read the updated version yet, so this may already be covered, but... we will need to support streams, (and soon tables), that have no KEY field set in the with clause.  In such a situation, how would the user set the key in the message?  We can't just say 'it will be null' as that will be too restrictive.  We need some syntax to allow the user to set it and we should keep in mind the discussions around the changes  in query semantics and structured key work to ensure any syntax is evolvable to support these.

Reply:
Yeah this discussion has had a lot of flip-flopping from myself. The most up-to-date one contradicts what I was saying here, and I think is more consistent (that is of course disregarding the future structured keys work)